apps/sms/services/js/moz_mobile_message/messaging_serivce.js

```diff
@@ -0,0 +1,176 @@
+        Settings.hasSeveralSim()) {
+      sendOpts = { serviceId: serviceId };
+    }
+
+    return sendOpts;
+  }
+
+  var MessagingService = {
+    /**
+     * Initializes our service using the Service mixin's initService and
+     * requests mozMobileMessage shim as client.
+     */
+    init() {
+      this.initService();
+
+      if (!('bridge' in self)) {
+        importScripts('/lib/bridge.js');
+      }
+      this.client = bridge.client(CLIENT_NAME);
+    },
+
+    /**
+     * Send sms by using send API shim.
+     * @param {*} opts Options for sending sms contains following properties:
+     * - recipients (string or array of string): contains the list of
+     *   recipients for this message.
+     * - content (string): the message's body.
+     * - serviceId (optional long or string): the SIM serviceId we use to send
+     *   the message.
+     * @returns {Promise} Promise with array of the request result.
+     */
+    sendSMS: function mm_send(opts) {
+      var recipients = opts.recipients || [],
+          content = opts.content,
+          serviceId = sanitizeServiceId(opts.serviceId);
+
+      if (!Array.isArray(recipients)) {
+        recipients = [recipients];
+      }
+
+      var i = 0;
+      var sendOpts = getSendOptionsFromServiceId(serviceId);
+      var requests = [];
+
+      recipients.forEach((recipient) => {
+        requests.push(this.client('send', recipient, content, sendOpts));
+      });
+
+      return Promise.all(requests).then((values) =>
+        values.map((value, idx) => ({
+          success: !event.target.error,
+          result: event.target.result,
+          code: event.target.error,
+          recipient: recipients[idx]
+        }));
+      );
+
+    },
+
+    /**
+     * Send mms by using sendMMS API shim.
+     * @param {*} opts Options for sending mms contains following properties:
+     * - recipients (string or array of string): recipients for this message.
+     * - subject (optional string): subject for this message.
+     * - content (array of SMIL slides): this is the content for the message (
+     *   see ConversationView for more information).
+     * - serviceId (optional long or string): the SIM that should be used for
+     *   sending this message. If this is not the current default configuration
+     *   for sending MMS, then we'll first switch the configuration to this
+     *   serviceId, and only then send the message. That means that the
+     *   "sending" event will come quite late in this case.
+     * @returns {Promise} DOMRequest promise with success result or error.
+     */
+    sendMMS: function mm_sendMMS(opts) {
+      var recipients = opts.recipients,
+          subject = opts.subject,
+          content = opts.content,
+          serviceId = opts.serviceId = sanitizeServiceId(opts.serviceId),
+
+      if (!Array.isArray(recipients)) {
+        recipients = [recipients];
+      }
+
+      var message = SMIL.generate(content);
```

Comment:
ideally generate and parse should be done in the same "layer" (or same object); but this could be a longer-term goal. I'm fine with a more pragmatic approach for now.